src/hotspot/share/utilities/debug.cpp

```diff
@@ -725,10 +725,22 @@ void disarm_assert_poison() {
 
 static void store_context(const void* context) {
   memcpy(&g_stored_assertion_context, context, sizeof(ucontext_t));
-#if defined(LINUX) && defined(PPC64)
+#if defined(LINUX)
   // on Linux ppc64, ucontext_t contains pointers into itself which have to be patched up
   //  after copying the context (see comment in sys/ucontext.h):
+#if defined(PPC64)
   *((void**) &g_stored_assertion_context.uc_mcontext.regs) = &(g_stored_assertion_context.uc_mcontext.gp_regs);
+#elif defined(AMD64)
+  // In the copied version, fpregs should point to the copied contents.
+  // Sanity check: fpregs should point into the context.
+  if ((address)((const ucontext_t*)context)->uc_mcontext.fpregs > (address)context) {
+    size_t fpregs_offset = pointer_delta(((const ucontext_t*)context)->uc_mcontext.fpregs, context, 1);
+    if (fpregs_offset < sizeof(ucontext_t)) {
```

Comment:
I think you should set `fpregs` to null if the original `fpregs` is invalid. The message in `print_context` will be confusing otherwise.
But I'll leave it to you assuming it'll never happen. Personally I think we should not try to avoid the assertion in `pointer_delta` and keep the code concise.

Reply:
I think null would be invalid, too. I prefer not touching it if it's already broken.